tls/extensions/s2n_client_key_share.c

```diff
@@ -98,6 +95,62 @@ static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s
     return S2N_SUCCESS;
 }
 
+static int s2n_send_hrr_keyshare(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+    const struct s2n_ecc_named_curve *server_negotiated_curve = NULL;
+    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
+
+    const struct s2n_ecc_preferences *ecc_pref = NULL;
+    GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
+    notnull_check(ecc_pref);
+
+    server_negotiated_curve = conn->secure.server_ecc_evp_params.negotiated_curve;
+    ENSURE_POSIX(server_negotiated_curve != NULL, S2N_ERR_BAD_KEY_SHARE);
+
+    bool has_supported_curve = false;
+    size_t supported_curve_index = 0;
+
+    for (size_t i = 0; i < ecc_pref->count; i++) {
+        if (ecc_pref->ecc_curves[i]->iana_id == server_negotiated_curve->iana_id) {
+            supported_curve_index = i;
+            has_supported_curve = true;
+            break;
+        }
+        /* Our original key shares weren't successful, so clear the old list of keyshares */
+        if (&conn->secure.client_ecc_evp_params[i] != NULL) {
+            GUARD(s2n_ecc_evp_params_free(&conn->secure.client_ecc_evp_params[i]));
+            conn->secure.client_ecc_evp_params[i].negotiated_curve = NULL;
+        }
+    }
+
+    ENSURE_POSIX(has_supported_curve, S2N_ERR_BAD_KEY_SHARE);
+
+    /* Generate the keyshare for the server negotiated curve */
+    ecc_evp_params = &conn->secure.client_ecc_evp_params[supported_curve_index];
+    ecc_evp_params->negotiated_curve = server_negotiated_curve;
+    ecc_evp_params->evp_pkey = NULL;
+    GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    /* From https://tools.ietf.org/html/rfc8446#section-4.1.2
+     * If a "key_share" extension was supplied in the HelloRetryRequest,
+     * replace the list of shares with a list containing a single
+     * KeyShareEntry from the indicated group.*/
+    if (s2n_is_hello_retry_handshake(conn)) {
+        GUARD(s2n_hello_retry_validate(conn));
```

Comment:
Is this s2n_hello_retry_validate actually required? Don't we always send the hrr keyshare once we've entered the hrr handshake? We're not in a hrr handshake for the first CH, and we are for the second CH.

Reply:
I added the check as a defensive mechanism, to verify the hello_retry is in a valid state.